Add support for tap to focus

justJS opened the pull request:
This pull request fixes #6 and adds support for tap to focus. As tap to focus isn't always preferred, tap to focus is only available in "Auto" mode, which has been renamed to reflect this.

This pull request:
- Adds support for tap to focus
- Add notification from `AVCaptureDevice` when the "subject area changed" so we can reset the focus to auto
- Adds the `FocusRectangleView` class, which displays the current tap to focus rectangle
- Renames `videoPreviewlayer` to `videoPreviewLayer` to reflect Swift naming standards

justJS commented on WeScan/Scan/ScannerViewController.swift:
```diff
@@ -177,19 +180,96 @@ final class ScannerViewController: UIViewController {
+                device.unlockForConfiguration()
+            } catch {
+                let error = ImageScannerControllerError.inputDevice
+                captureSessionManager?.delegate?.captureSessionManager(captureSessionManager!, didFailWithError: error)
```
Resolved: 8f2e8e7

justJS commented on WeScan/Scan/ScannerViewController.swift:
```diff
@@ -177,19 +180,96 @@ final class ScannerViewController: UIViewController {
+            device.unlockForConfiguration()
+        } catch {
+            let error = ImageScannerControllerError.inputDevice
+            captureSessionManager?.delegate?.captureSessionManager(captureSessionManager!, didFailWithError: error)
```
Resolved: 8f2e8e7

justJS commented on WeScan/Scan/FocusRectangle.swift:
```diff
@@ -0,0 +1,32 @@
+
+import UIKit
+
+class FocusRectangle: UIView {
```
Resolved: 8f2e8e7

justJS commented:
Note: I might not be available over the next ~week or so to fix issues, but I'll keep working on this when I'm back.

jcampbell05 commented on WeScan/Common/CaptureSession.swift:
```diff
@@ -17,14 +17,14 @@ final class CaptureSession {
     
     /// Whether auto scan is enabled or not
-    var autoScanEnabled: Bool
+    var isAutoModeEnabled: Bool
```
We don't mention tap to focus in the documentation for this variable, I think if we are going to go with this flag we should explain when they want this to be false so it's clear how to opt into the tap to focus behavior.

justJS replied:
Sure, will do.

justJS replied:
Resolved: 9d1e1e4

jcampbell05 commented on WeScan/Scan/FocusRectangle.swift:
```diff
@@ -0,0 +1,32 @@
+
+import UIKit
+
+final class FocusRectangle: UIView {
```
Can we add a line of text to explain that this is for the tap to focus UI like you have in this PR

justJS replied:
Resolved: 9d1e1e4

jcampbell05 commented on WeScan/Scan/FocusRectangle.swift:
```diff
@@ -0,0 +1,32 @@
+        layer.borderWidth = 2.0
+        layer.cornerRadius = 6.0
+        layer.borderColor = UIColor.yellow.cgColor
+        
```
There are a lot of magic numbers and calculations heres, is there a way of breaking this down or putting some of these into variables to explain whats going on ?

justJS replied:
Resolved: 9d1e1e4

jcampbell05 commented on WeScan/Scan/ScannerViewController.swift:
```diff
@@ -177,19 +180,98 @@ final class ScannerViewController: UIViewController {
+            captureSessionManager.delegate?.captureSessionManager(captureSessionManager, didFailWithError: error)
+            return
+        }
+        
```
I think this function is getting a bit long, it would be nice to split this up into smaller functions that this one calls. Preferably with nice names so we can understand the steps taken.

justJS replied:
Resolved: 9d1e1e4

jcampbell05 commented on WeScan/Scan/ScannerViewController.swift:
```diff
@@ -177,19 +180,98 @@ final class ScannerViewController: UIViewController {
+                    focusRectangle.alpha = 0.0
+                }, completion: { (_) in
+                    focusRectangle.removeFromSuperview()
+                })
```
Again I think we can make this one shorter.

justJS replied:
Resolved: 9d1e1e4

jcampbell05 commented:
Looks like a useful addition. I've added some comments on some bits I think we can breakdown to make more understandable for others.

Once we've done that I think we can review the code, right now I'm having a hard time understanding what each line does :).

justJS commented:
Done @jcampbell05 😄

AvdLee commented on WeScan/Common/CaptureSession.swift:
```diff
@@ -16,15 +16,15 @@ final class CaptureSession {
-    /// Whether auto scan is enabled or not
-    var autoScanEnabled: Bool
+    /// The status of "auto mode". Auto mode does not allow tap to focus and tries to automatically scan a detected rectangle if it has a high enough accuracy.
+    var isAutoModeEnabled: Bool
```
Renaming this to `isAutoFocusEnabled` would improve the understanding of the variable

justJS replied:
Hey, as per the comment auto mode is actually both auto scan and auto focus, so renaming it to `isAutoFocusEnabled` wouldn't make sense IMO.

I did consider separating the 2 modes, but I felt it would be difficult to represent what each icon meant on the limited space available on the toolbar (on the scanner screen).

justJS replied:
Renamed to isAutoScanEnabled as auto mode no longer matters for auto focus.

AvdLee commented on WeScan/Scan/ScannerViewController.swift:
```diff
@@ -54,8 +57,8 @@ final class ScannerViewController: UIViewController {
     
-    lazy private var autoScanButton: UIBarButtonItem = {
-        return UIBarButtonItem(title: NSLocalizedString("wescan.scanning.auto", tableName: nil, bundle: Bundle(for: ScannerViewController.self), value: "Auto", comment: "The auto button state"), style: .plain, target: self, action: #selector(toggleAutoScan))
+    lazy private var autoModeButton: UIBarButtonItem = {
```
`autoMode` sounds pretty vague to me, especially without documentation. Can we make this property more descriptive and can we add some documentation?

justJS replied:
Hey, autoMode is documented on the `isAutoModeEnabled` property of the `CaptureSession` 😃

AvdLee commented on WeScan/Scan/ScannerViewController.swift:
```diff
@@ -175,6 +178,82 @@ final class ScannerViewController: UIViewController {
+            captureSessionManager.delegate?.captureSessionManager(captureSessionManager, didFailWithError: error)
+            return
+        }
+    }
```
We're adding quite some logic to the `ScannerViewController`, making it responsible for more and more things. Would be great if we can migrate this part of the code to a `ScanFocusController` something like class, keeping this `ScannerViewController` smaller. 

That should also make it easier to test the code.

justJS replied:
Done! 399be14

justJS commented:
Hey @AvdLee, addressed some of your comments above and will keep working on the logic separation.

As per the unit tests, I feel there isn't much that we can do. `isAutoModeEnabled` is already tested in the `CaptureSession` tests, and actually enabling/disabling autoFocus on the `AVCaptureVideoPreviewLayer` is handled by the system. Did you have any specific ideas for tests in mind?

AvdLee commented on WeScan/Common/CaptureSession.swift:
```diff
@@ -29,3 +30,90 @@ final class CaptureSession {
+            focusRectangle.removeFromSuperview()
+        }
+    }
+}
```
This part should be testable and was what I was aiming for. Should be doable with protocols. Let me know if you have any questions!

justJS replied:
Hey, not really sure how you intend this to be tested? Do you just want a ScannerVC to be created, then these methods called and stuff like the focus rectangle existing/not existing being tested?

jcampbell05 replied:
I think the easiest way to test this is just to create a CaptureSession instance and test that when this is called the focusRectangle is not longer in the view hierarchy.

Unfortunately the animation will be hard to test unless you disable animations gloaballly using CATranasaction so I think we should just agree to test the non-animated variant for now.

jcampbell05 replied:
Hopefully this can help you see ways you can make the function to be easier to test once you've written the first version of those tests :)

justJS commented:
Hey all, I'm thinking about closing this PR. 
While I do think tap to focus could possibly be useful, auto focus is typically good enough, and I feel that at least our current implementation misses a few things, namely it's hard to discover (only available in Manual mode) and it's not obvious what "Auto" mode does anymore, while when it was just auto scan, it was way more obvious.

Any thoughts on this?

justJS commented:
Sounds like a good idea @Boris-Em! I've researched and there does seem to be a property to receive notifications if the "content area changes" (from AVCaptureDevice). I'll keep working on this when I have a bit more time 🙂

Boris-Em commented:
Great! Thank you so much @justJS!

AvdLee commented on WeScan/Common/CaptureSession.swift:
```diff
@@ -16,16 +17,103 @@ final class CaptureSession {
+    /// Sets the camera's exposure and focus point to the given point
+    func setFocusPointToTapPoint(_ tapPoint: CGPoint) throws {
+        guard let device = AVCaptureDevice.default(for: AVMediaType.video) else { return }
+        try device.lockForConfiguration()
```
We can use `defer` here as well for consistency

AvdLee commented on WeScan/Common/CaptureSession.swift:
```diff
@@ -16,16 +17,103 @@ final class CaptureSession {
+    /// Resets the camera's exposure and focus point to automatic
+    func resetFocusToAuto() throws {
+        guard let device = AVCaptureDevice.default(for: AVMediaType.video) else { return }
+        try device.lockForConfiguration()
```
We can use `defer` here as well for consistency

justJS commented:
Thanks! Any specific code you want to see tested? Stuff like focus setting is handled by AVCaptureDevice, not sure how to test it?

AvdLee commented:
@justJS if you take a look at the Danger message:

## Current coverage for WeScan is `34.35%`
Files changed | - | - 
--- | --- | ---
ScannerViewController.swift | `0.00%` | :skull:
CaptureSessionManager.swift | `0.00%` | :skull:
CaptureSession.swift | `14.63%` | :skull:
FocusRectangleView.swift | `100.00%` | :white_check_mark:

---

You can see that quite some files are far from having a nice code coverage. Our aim is to have around 80%. We can't reach that with this PR only, of course, but if you can add tests for the methods you've added in these classes, we at least get a bit closer 💪 

Classes like the `AVCaptureDevice` can be tested by using protocols. See this blog post for example for inspiration: http://masilotti.com/testing-nsurlsession-input/

justJS commented:
Thanks for the pointers! I’ll work on this later and see what I can do 😃

justJS commented:
Resolved merge conflicts from the merge earlier today of #83. Will work on tests tomorrow/day after 🙂

justJS commented:
Added tests for flash and focus based on a MockCaptureDevice! @AvdLee @Boris-Em 

## Current coverage for WeScan is `38.36%`
Files changed | - | -
--- | --- | ---
CaptureSession+Focus.swift | `75.81%` | :warning:
CaptureSession+Flash.swift | `78.26%` | :warning:
CaptureSession.swift | `100.00%` | :white_check_mark:
CaptureDevice.swift | `100.00%` | :white_check_mark:
FocusRectangleView.swift | `100.00%` | :white_check_mark:

justJS commented:
Fixed the merge conflict that occurred by merging #60 🙂

justJS commented:
Are we waiting for any further reviews here?

AvdLee commented:
Added @Boris-Em to review this and test on an actual device. After that, it's good!
